docs/general/cairo_precompiles.md

```diff
@@ -0,0 +1,170 @@
+# Cairo Precompiles
+
+Kakarot zkEVM being a Starknet appchain, it is technically possible to run Cairo
+Contracts on Kakarot. The purpose of this document is to explain the design
+behind the Cairo precompiles, which are the Cairo contracts that are deployed on
+Kakarot to provide additional functionality to the users.
+
+## Requirements
+
+From a developer / user perspective, the precompiles should be easy to interact
+with from EVM contracts. As such, we need to provide a way to call the
+precompiles from EVM contracts, and to return the results to the EVM contracts.
+This involves:
+
+- Converting the EVM calldata to Cairo inputs
+- Converting the Cairo outputs to EVM return data
+
+From a security perspective, the precompiles should _never_ produce Cairo errors
+that would cause the transaction to revert. As we cannot catch contract call
+errors in Cairo, we need to ensure that executing a `call_contract` syscall to a
+precompile will never cause the transaction to revert, meaning that:
+
+- The target Cairo addresses should always correspond to a deployed contract
+- The selector of the Cairo function being called should always be present in
+  the contract called
+- The Cairo contract called should never _panic_
+
+From these principles, we can derive the following design.
+
+## Design
+
+Interacting with the Cairo precompiles will only be done by calling specific EVM
+contracts that have been whitelisted for this purpose. This ensures that the
+Cairo precompiles are only called by contracts that have been reviewed and
+approved regarding the security concerns mentioned above.
+
+As such, the execution flow of an EVM message will be as follows:
+
+```mermaid
+flowchart TD
+    A[Solidity contract] -->|ICairo.call_contract| B
+    B{address type}
+    B -->|EVM Contract| D[Execute contract]
+    B -->|EVM Precompiles| E[Execute EVM precompile]
+    B -->|Cairo Precompile| F{Whitelisted caller code ?}
```

Comment:
whitelisted starknet address and entrypoint

Reply:
Whitelist happens at the solidity level, not starknet 🤔 (or both?)

Reply:
I think that we whitelist starknet contract + its entrypoints, not EVM caller. Once a precompile (starknet contract) is deemed legit, why not letting anyone else calling it ? because also a wrong calldata can trigger a VM error ?

Reply:
in cairo1, `call_contract` syscall can fail safe right? so maybe we could add in the `Cairo1Helper` a function to actually do the `call_contract` with the new syscall, instead of the whitelist mechanism?

Reply:
@ClementWalter no, call_contract cannot fail safe. Ariel told me about the possibility to handle syscall errors for release 0.15 (EOY at best)